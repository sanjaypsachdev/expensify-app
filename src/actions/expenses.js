import realtimeDb, { expensesFs } from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => {
  return ({
  type: 'ADD_EXPENSE',
  expense
})};

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return expensesFs(uid)
      .add(expense)
      .then((docRef) => {
        dispatch(addExpense({
          id: docRef.id,
          ...expense
        }));
      });

    // return realtimeDb.ref(`users/${uid}/expenses`)
    //   .push(expense)
    //   .then((ref) => {
    //     dispatch(addExpense({
    //       id: ref.id,
    //       ...expense
    //     }));
    //   });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    //return realtimeDb.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
    return expensesFs(uid).doc(id).delete().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    //return realtimeDb.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
    return expensesFs(uid).doc(id).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    //return realtimeDb.ref(`users/${uid}/expenses).once('value').then((snapshot) => {
    return expensesFs(uid).get().then((snapshot) => {
      const expenses = [];

      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.id,
          ...childSnapshot.data()
        });
      });

      dispatch(setExpenses(expenses));
    });
  };
};
