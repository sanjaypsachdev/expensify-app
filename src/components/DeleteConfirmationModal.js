import React from 'react';
import ReactModal from 'react-modal';
import { connect } from 'react-redux';
import {
  startRemoveExpense,
  cancelRemoveExpense } from '../actions/expenses';

export class DeleteConfirmationModal extends React.Component {

  onRemoveExpense = () => {
    this.props.startRemoveExpense({ id: this.props.expenseToDelete.id });
    this.props.afterRemoveExpense();
  }
  
  handleCancelRemoveExpense = () => {
    this.props.cancelRemoveExpense({ id: this.props.expenseToDelete.id })
  }
  
  render() {
    return(
      <ReactModal
        ariaHideApp={false}
        isOpen={this.props.expenseToDelete.isDeleting}
        onRequestClose={this.handleCancelRemoveExpense}
        contentLabel="Confirm Deletion"
        className="modal"
      >
        <h3 className="modal__title">Are you sure you want to delete this expense ?</h3>
        <div className="modal_buttons">
          <button className="button yes" onClick={this.onRemoveExpense}>Yes</button>
          <button className="button no" onClick={this.handleCancelRemoveExpense}>No</button>
        </div>
      </ReactModal>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
  cancelRemoveExpense: (id) => dispatch(cancelRemoveExpense(id))
});

export default connect(null, mapDispatchToProps)(DeleteConfirmationModal);