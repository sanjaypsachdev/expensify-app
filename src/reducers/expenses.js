// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'SET_REMOVE_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            isDeleting: true
          };
        } else {
          return expense;
        };
      });
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'CANCEL_REMOVE_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            isDeleting: undefined
          };
        } else {
          return expense;
        };
      });
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        };
      });
    case 'SET_EXPENSES':
      return action.expenses;
    case 'CLEAR_ALL_EXPENSES':
      return [];
    default:
      return state;
  }
};
