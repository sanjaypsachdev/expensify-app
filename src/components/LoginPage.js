import React from 'react';
import { connect } from 'react-redux';
import { startLogin, authProviders } from '../actions/auth';

export const LoginPage = ({ startGoogleLogin, startGithubLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>It's time to get your expenses under control.</p>
      <div className="box-layout__buttons">
        <button className="button" onClick={startGoogleLogin}>Login with Google</button>
        <button className="button" onClick={startGithubLogin}>Login with Github</button>
      </div>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startGoogleLogin: () => dispatch(startLogin(authProviders.GOOGLE)),
  startGithubLogin: () => dispatch(startLogin(authProviders.GITHUB))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);