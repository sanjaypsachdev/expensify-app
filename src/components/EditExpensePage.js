import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {
  startEditExpense,
  setRemoveExpense } from '../actions/expenses';
import DeleteConfirmationModal from './DeleteConfirmationModal';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  onSetRemoveExpense = () => {
    this.props.setRemoveExpense({ id: this.props.expense.id })
  }

  afterRemoveExpense = () => {
    this.props.history.push('/');
  }  

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onSetRemoveExpense}>Remove Expense</button>
        </div>
        <DeleteConfirmationModal
          expenseToDelete={this.props.expense}
          afterRemoveExpense={this.afterRemoveExpense}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (expenseId, expense) => dispatch(startEditExpense(expenseId, expense)),
  setRemoveExpense: (id) => dispatch(setRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
