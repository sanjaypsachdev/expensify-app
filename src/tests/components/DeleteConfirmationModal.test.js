import React from 'react';
import ReactModal from 'react-modal';
import { shallow } from 'enzyme';
import { DeleteConfirmationModal } from '../../components/DeleteConfirmationModal';
import expenses from '../fixtures/expenses';

let startRemoveExpense, cancelRemoveExpense, afterRemoveExpense, wrapper;

beforeEach(() => {
  startRemoveExpense = jest.fn();
  cancelRemoveExpense = jest.fn();
  afterRemoveExpense = jest.fn();

  const expenseToDelete = { ...expenses[0], isDeleting: true }
  wrapper = shallow(<DeleteConfirmationModal
    startRemoveExpense={startRemoveExpense}
    cancelRemoveExpense={cancelRemoveExpense}
    afterRemoveExpense={afterRemoveExpense}
    expenseToDelete={expenseToDelete}
  />);
});

test('should render DeleteConfirmationModal correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should be open', () => {
  expect(wrapper.find(ReactModal).prop('isOpen')).toEqual(true);
})

test('should handle start remove expense', () => {
  wrapper.find('.yes').prop('onClick')();
  expect(afterRemoveExpense).toHaveBeenCalled();
  const id = expenses[0].id;
  expect(startRemoveExpense).toHaveBeenLastCalledWith({ id });
});

test('should handle cancel remove expense', () => {
  wrapper.find('.no').prop('onClick')();
  const id = expenses[0].id;
  expect(cancelRemoveExpense).toHaveBeenLastCalledWith({ id });
});