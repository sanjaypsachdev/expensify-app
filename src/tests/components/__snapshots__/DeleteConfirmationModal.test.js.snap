// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DeleteConfirmationModal correctly 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="Confirm Deletion"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Are you sure you want to delete this expense ?
  </h3>
  <div
    className="modal_buttons"
  >
    <button
      className="button yes"
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="button no"
      onClick={[Function]}
    >
      No
    </button>
  </div>
</Modal>
`;
